engines/wage/design.cpp

```diff
@@ -0,0 +1,534 @@
+	Common::Rect inn(x1-5, y1-5, x2+5, y2+5);
+	drawRoundRect(inn, 6, kColorGray, false, drawPixelPlain, &pd);
+
+	drawThickLine(x1, y1, x2-borderThickness, y1, borderThickness, kColorBlack, drawPixel, &pd);
+	drawThickLine(x2-borderThickness, y1, x2-borderThickness, y2, borderThickness, kColorBlack, drawPixel, &pd);
+	drawThickLine(x2-borderThickness, y2-borderThickness, x1, y2-borderThickness, borderThickness, kColorBlack, drawPixel, &pd);
+	drawThickLine(x1, y2-borderThickness, x1, y1, borderThickness, kColorBlack, drawPixel, &pd);
+	drawThickLine(x2+10, y2+10, x2+100, y2+100, borderThickness, kColorBlack, drawPixel, &pd);
+
+	g_system->copyRectToScreen(_surface->getPixels(), _surface->pitch, 0, 0, _surface->w, _surface->h);
+
+	while (true) {
+		((WageEngine *)g_engine)->processEvents();
+		g_system->updateScreen();
+		g_system->delayMillis(50);
+	}
+	return;
+#endif
+
+	while (needRender) {
+		byte fillType = in.readByte();
+		byte borderThickness = in.readByte();
+		byte borderFillType = in.readByte();
+		int type = in.readByte();
+
+		if (in.eos())
+			break;
+
+		debug(8, "fill: %d borderFill: %d border: %d type: %d", fillType, borderFillType, borderThickness, type);
+		switch (type) {
+		case 4:
+			drawRect(_surface, in, patterns, fillType, borderThickness, borderFillType);
+			break;
+		case 8:
+			drawRoundRect(_surface, in, patterns, fillType, borderThickness, borderFillType);
+			break;
+		case 12:
+			drawOval(_surface, in, patterns, fillType, borderThickness, borderFillType);
+			break;
+		case 16:
+		case 20:
+			drawPolygon(_surface, in, patterns, fillType, borderThickness, borderFillType);
+			break;
+		case 24:
+			drawBitmap(_surface, in);
+			break;
+		default:
+			warning("Unknown type => %d", type);
+			break;
+		}
+
+		//g_system->copyRectToScreen(_surface->getPixels(), _surface->pitch, 0, 0, _surface->w, _surface->h);
+		//((WageEngine *)g_engine)->processEvents();
+		//g_system->updateScreen();
+		//g_system->delayMillis(500);
+	}
+
+	const int padding = 3;
+	for (int i = padding; i < _bounds->height() - 2 * padding; i++) {
+		const byte *src = (const byte *)_surface->getBasePtr(padding, i);
+		byte *dst = (byte *)surface->getBasePtr(x + padding, y+i);
+		for (int j = padding; j < _bounds->width() - 2 * padding; j++) {
+			if (*src != kColorGreen)
+				*dst = *src;
+			src++;
+			dst++;
+		}
+	}
+}
+
+bool Design::isPointOpaque(int x, int y) {
+	if (_surface == NULL)
+		error("Surface is null");
+
+	byte pixel = ((byte *)_surface->getBasePtr(x, y))[0];
+
+	return pixel != kColorGreen;
+}
+
+void drawPixel(int x, int y, int color, void *data) {
+	plotData *p = (plotData *)data;
+
+	if (p->fillType > p->patterns->size())
+		return;
+
+	byte *pat = p->patterns->operator[](p->fillType - 1);
+
+	if (p->thickness == 1) {
+		if (x >= 0 && x < p->surface->w && y >= 0 && y < p->surface->h) {
+			*((byte *)p->surface->getBasePtr(x, y)) =
+				(pat[y % 8] & (1 << (7 - x % 8))) ?
```

Comment:
These %8's will produce suboptimal code if optimizations are disabled since x and y are signed.